internal/validation/validation.go

```diff
@@ -166,6 +170,93 @@ func validateTypeRestrictions(typesys *typesystem.TypeSystem, tk *openfgav1.Tupl
 	return fmt.Errorf("type '%s' is not an allowed type restriction for '%s#%s'", userType, objectType, tk.GetRelation())
 }
 
+// validateCondition enforces conditions on a relationship tuple
+func validateCondition(typesys *typesystem.TypeSystem, tk *openfgav1.TupleKey) error {
+	objectType := tuple.GetType(tk.Object)
+	userType := tuple.GetType(tk.User)
+	userRelation := tuple.GetRelation(tk.User)
+
+	typeRestrictions, err := typesys.GetDirectlyRelatedUserTypes(objectType, tk.Relation)
+	if err != nil {
+		return err
+	}
+
+	if tk.Condition == nil {
+		hasConditionedTypeRestriction := false
+		hasUnconditionedTypeRestriction := false
+
+		for _, userset := range typeRestrictions {
+			if userset.Type != userType {
+				continue
+			}
+
+			if userset.Condition == "" {
+				hasUnconditionedTypeRestriction = true
+				continue
+			}
+
+			if userset.RelationOrWildcard == nil {
+				hasConditionedTypeRestriction = true
+				continue
+			}
+
+			if _, ok := userset.RelationOrWildcard.(*openfgav1.RelationReference_Relation); ok {
+				if userset.GetRelation() == userRelation {
+					hasConditionedTypeRestriction = true
+					continue
+				}
+			}
+			if _, ok := userset.RelationOrWildcard.(*openfgav1.RelationReference_Wildcard); ok {
+				hasConditionedTypeRestriction = true
+				continue
+			}
+		}
+
+		// A condition is only required if all the direct relationship type
+		// restrictions are conditioned. Example:
+		// define viewer: [user, user with condition]: not required
+		// define viewer: [user]: not required
+		// define viewer: [user with condition]: required
+		if hasConditionedTypeRestriction && !hasUnconditionedTypeRestriction {
+			return &tuple.InvalidConditionalTupleError{
+				Cause: fmt.Errorf("condition is missing"), TupleKey: tk,
+			}
+		}
+
+		return nil
+	} else {
+		condition, ok := typesys.GetConditions()[tk.Condition.Name]
+		if !ok {
+			return &tuple.InvalidConditionalTupleError{
+				Cause: fmt.Errorf("undefined condition"), TupleKey: tk,
+			}
+		}
+
+		validCondition := false
+		for _, userset := range typeRestrictions {
+			if userset.Type == userType && userset.Condition == tk.Condition.Name {
+				validCondition = true
+				continue
```

Comment:
No need to `continue`, we can break out of the containing loop, because we found at least 1 type restriction that allows the relationship with the condition.